File: 0347-top-k-frequent-elements/0347-top-k-frequent-elements.cpp
class Solution {
public:
    vector<int> topKFrequent(vector<int>& nums, int k)
    {
        unordered_map<int,int> map;
        
        for(auto & num : nums)
        {
            if(map.find(num) == map.end())
            {
                map[num] = 1;
            }
            else
            {
                map[num]++;
            }
        }
        
        vector<vector<int>> candidates;
        
        for(auto& [key, val] : map)
        {
            candidates.push_back({val, key});
        }
        
        return quickselect(0, candidates.size() - 1, k, candidates);
    }
    
    vector<int> quickselect(int left, int right, int k, vector<vector<int>> & candidates)
    {
        
        // S S S S S O O O L L L L 
        //           i t j
        
        if(left == right)
        {
            return {candidates[left][1]};
        }
        
        int mid = (right - left) / 2 + left;
        
        int pivot = candidates[mid][0];
        
        int i = left, j = right, t = left;
        
        while(t <= j)
        {
            if(candidates[t][0] < pivot)
            {
                vector<int> temp = candidates[t];
                candidates[t] = candidates[i];
                candidates[i] = temp;
                i++;
                t++;
            }
            else if(candidates[t][0] > pivot)
            {
                vector<int> temp = candidates[t];
                candidates[t] = candidates[j];
                candidates[j] = temp;
                j--;
            }
            else
            {
                t++;
            }
        }
        
        vector<int> ans;
        
        if(right - i + 1 == k)
        {
            for(int p = i; p <= right; p++)
            {
                ans.push_back(candidates[p][1]);
            }
            
            return ans;
        }
        else if(right - i + 1 < k)
        {
            for(int p = i; p <= right; p++)
            {
                ans.push_back(candidates[p][1]);
            }
            
            vector<int> rights = quickselect(left, i - 1, k - (right - i + 1), candidates);
            
            for(auto & num : rights)
            {
                ans.push_back(num);
            }
            
            return ans;
        }
        else
        {
            return quickselect(i + 1, right, k, candidates);
        }
    }
};

File: 0347-top-k-frequent-elements/0347-top-k-frequent-elements.py
class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        hashmap = {}
        
        for num in nums:
            hashmap[num] = hashmap.get(num, 0) + 1
        
        candidates = []
        
        for key, val in hashmap.items():
            candidates.append([key,val])
            
            
        def quickselect(l, r, k):
            if l == r:
                return [candidates[l][0]]
            
            mid = (r - l) // 2 + l
            pivot = candidates[mid][1]
            
            i, t, j = l, l, r
            
            while t <= j:
                if candidates[t][1] < pivot:
                    candidates[t], candidates[i] = candidates[i], candidates[t]
                    i += 1
                    t += 1
                elif candidates[t][1] > pivot:
                    candidates[t], candidates[j] = candidates[j], candidates[t]
                    j -= 1
                else:
                    t += 1
            
            # S S S S E E E E L L L L
            
            if r - j + 1 == k:
                return [x[0] for x in candidates[j:r + 1]]
            elif r - j + 1 > k:
                return quickselect(j + 1, r, k)
            else:
                return [x[0] for x in candidates[j:r + 1]] + quickselect(l, j - 1, k - (r - j + 1))
            
        return quickselect(0, len(candidates) - 1, k)